network/src/peer_store/peer_store_impl.rs

```diff
@@ -0,0 +1,244 @@
+                    format!("report behaviour {:?}", behaviour),
+                )?;
+                return Ok(ReportResult::Banned);
+            }
+        }
+        Ok(ReportResult::Ok)
+    }
+
+    pub fn remove_disconnected_peer(&mut self, peer_id: &PeerId) -> Option<PeerInfo> {
+        self.peers.borrow_mut().remove(peer_id)
+    }
+
+    pub fn peer_status(&self, peer_id: &PeerId) -> Status {
+        if self.peers.borrow().contains_key(peer_id) {
+            Status::Connected
+        } else {
+            Status::Disconnected
+        }
+    }
+
+    /// Get peers for outbound connection, this method randomly return non-connected peer addrs
+    pub fn get_addrs_to_attempt(&mut self, count: usize) -> Vec<AddrInfo> {
+        let now_ms = faketime::unix_time_as_millis();
+        let ban_list = self.ban_list.borrow();
+        let peers = self.peers.borrow();
+        // get addrs that can attempt.
+        self.addr_manager.get_random(count, |peer_addr: &AddrInfo| {
+            !ban_list.is_addr_banned(&peer_addr.addr)
+                && !peers.contains_key(&peer_addr.peer_id)
+                && !peer_addr.tried_in_last_minute(now_ms)
+        })
+    }
+
+    /// Get peers for feeler connection, this method randomly return peer addrs that we never
+    /// connected to.
+    pub fn get_addrs_to_feeler(&mut self, count: usize) -> Vec<AddrInfo> {
+        let now_ms = faketime::unix_time_as_millis();
+        let addr_expired_ms = now_ms - ADDR_TIMEOUT_MS;
+        // get expired or never successed addrs.
+        let ban_list = self.ban_list.borrow();
+        let peers = self.peers.borrow();
+        self.addr_manager.get_random(count, |peer_addr: &AddrInfo| {
+            !ban_list.is_addr_banned(&peer_addr.addr)
+                && !peers.contains_key(&peer_addr.peer_id)
+                && !peer_addr.tried_in_last_minute(now_ms)
+                && !peer_addr.had_connected(addr_expired_ms)
+        })
+    }
+
+    /// return valid addrs that success connected, used for discovery.
+    pub fn get_random_addrs(&mut self, count: usize) -> Vec<AddrInfo> {
+        let now_ms = faketime::unix_time_as_millis();
+        let addr_expired_ms = now_ms - ADDR_TIMEOUT_MS;
+        let ban_list = self.ban_list.get_mut();
+        // get success connected addrs.
+        self.addr_manager.get_random(count, |peer_addr: &AddrInfo| {
+            !ban_list.is_addr_banned(&peer_addr.addr) && peer_addr.had_connected(addr_expired_ms)
+        })
+    }
+
+    /// Ban a addr
+    pub(crate) fn ban_addr(
+        &mut self,
+        addr: &Multiaddr,
+        timeout_ms: u64,
+        ban_reason: String,
+    ) -> Result<()> {
+        let network = ip_to_network(addr.extract_ip_addr()?.ip);
+        self.ban_network(network, timeout_ms, ban_reason)
+    }
+
+    pub(crate) fn ban_network(
+        &mut self,
+        network: IpNetwork,
+        timeout_ms: u64,
+        ban_reason: String,
```

Comment:
Suggest to use `&str`
```suggestion
        ban_reason: &str,
```

Reply:
`String` is better, we can't use `format!` with `&str`, it's limit the reason report.

Reply:
You can `format!("{} xxx", yyy).as_str()`. But I think it's better use `String` here, since we take the ownership to `BannedAddr`.

Reply:
I have no strong opinions one way or the other, only concerned with consistency `NetworkState#ban` 
```
pub fn ban(&self, address: IpNetwork, ban_until: u64, ban_reason: &str)
```